### DISCUSSION TITLE
Use CI/CD Variables on composer update

### DISCUSSION THREAD

**Discussion by JanVormeister** [renovatebot/renovate#25348] [Request Help]
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

36.37.0

### If you're self-hosting Renovate, select which platform you are using.

GitLab self-hosted

### Was this something which used to work for you, and then stopped?

I never saw this working

### Wanted end result.

When running `composer update` through Renovat Runner, Renovate should have access to the GitLab CI/CD Variables.

Example composer.json::
```
...
"repositories": [
    {
      "type": "package",
      "package": {
        "name": "advanced-custom-fields/advanced-custom-fields-pro",
        "version": "6.0.0",
        "type": "wordpress-plugin",
        "dist": {
          "type": "zip",
          "url": "https://connect.advancedcustomfields.com/index.php?a=download&p=pro&k={%ACF_PRO_KEY}&t={%VERSION}"
        },
        "require": {
          "ffraenz/private-composer-installer": "^5.0"
        }
      }
    }
  ],
...
```

Here you can see the URL uses Placeholder which getting filled by CI/CD Variables from GitLab. But the update process fails, because the URL is, obviously, wrong. 
Renovate should not look up any versions. Instead simply install the package.

### What you tried so far.

I tried customEnvVariables, exposeAllEnv. 

### Relevant debug logs

<details><summary>Logs</summary>

```
Copy/paste the relevant log(s) here, between the starting and ending backticks
```

</details>

**Comment by rarkins** [comment]
`exposeAllEnv` should work as described. If composer can't see the env you need, it might mean that they're not being passed through to the Renovate process/container.

**Reply by JanVormeister** [reply]
Thanks @rarkins for your quick answer! 
Is there anything I have to set up to pass the env to the process/container, either in the renovate-runner or in the repository config?

**Reply by rarkins** [reply]
You need to provide much more details about how you run Renovate if I have any chance of guessing a problem in your environment

**Reply by JanVormeister** [reply]
Alright, I did not set up renovate for us, so I can just give you the information I can see, so I'll try my best :)

- As far as I can see, we use this runner template: https://gitlab.com/renovate-bot/renovate-runner/-/raw/v15.16.0/templates/renovate.gitlab-ci.yml
- A Renovate Token is set and a GitHub Token is set
- as extra flag is only autodiscovery set to true, nothing else

This is the current repository renovate.json
```
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "extends": [
    "config:recommended",
    "schedule:earlyMondays"
  ],
  "platform": "gitlab",
  "includePaths": [
    "composer.json",
    "web/app/themes/child-theme/**"
  ],
  "rangeStrategy": "update-lockfile",
  "branchPrefix": "deps/",
  "ignoreDeps": [
    "advanced-custom-fields/advanced-custom-fields-pro"
  ],
  "packageRules": [
    {
      "description": "Group dependencies from all files",
      "matchPackagePatterns": [
        "*"
      ],
      "matchUpdateTypes": [
        "major",
        "minor",
        "patch"
      ],
      "groupName": "All changes"
    }
  ],
  "separateMajorMinor": false,
  "githubTokenWarn": false
}
```

The lookup and version bump runs without issues. Renovate fails when running `composer update` with the message: `The \"https://connect.advancedcustomfields.com/index.php?a=download&p=pro&k=  \n  {%ACF_PRO_KEY}&t={%VERSION}\" file could not be downloaded (HTTP/2 404 ):     \n  {\"code\":\"invalid_version\",\"message\":\"No download was found for this version  \n  \",\"data\":{\"status\":404}}` 


Do you need any further information?

**Reply by rarkins** [reply]
Does Composer normally expand env into `{%ACF_PRO_KEY}` type variables?

Where/how do you set exposeAllEnv=true?

**Reply by JanVormeister** [reply]
Yes, within our GitLab pipeline composer can read the CI/CD variable `ACF_PRO_KEY`, locally it reads it from the `.env`.

I set the renovate env variable like this in the runner repo as well as the project repo: 
![grafik](https://github.com/renovatebot/renovate/assets/80897586/829a1b2d-83c5-40ec-95ba-c8fa367d6318)

**Reply by rarkins** [reply]
Please copy/paste the Renovate startup log to confirm it's being read as intended

**Reply by JanVormeister** [reply]
```
Running with gitlab-runner 16.4.0 (4e724e03)
  on gitlab-runner01.fsn1-dc8.infra.yay.digital c0a90e76, system ID: s_09cd4f57780e
Preparing the "docker" executor 00:42
Using Docker executor with image ghcr.io/renovatebot/renovate:36.37.0@sha256:{{MASKED FOR GITHUB DISKUSSION} ...
Pulling docker image ghcr.io/renovatebot/renovate:36.37.0@sha256:{{MASKED FOR GITHUB DISKUSSION} ...
Using docker image sha256:{{MASKED FOR GITHUB DISKUSSION}} for ghcr.io/renovatebot/renovate:36.37.0@sha256:{{MASKED FOR GITHUB DISKUSSION} with digest ghcr.io/renovatebot/renovate@sha256:{{MASKED FOR GITHUB DISKUSSION} ...
Preparing environment 00:01
Running on runner-c0a90e76-project-493-concurrent-0 via gitlab-runner01...
Getting source from Git repository 00:01
Fetching changes with git depth set to 20...
Initialized empty Git repository in /builds/{{MASKED FOR GITHUB DISKUSSION}/renovate-runner/.git/
Created fresh repository.
Checking out 50c83119 as detached HEAD (ref is main)...
Skipping Git submodules setup
Restoring cache 00:01
Checking cache for main-renovate-3-protected...
No URL provided, cache will not be downloaded from shared cache server. Instead a local version of cache will be extracted. 
Successfully extracted cache
Executing "step_script" stage of the job script 03:51
Using docker image sha256:{{MASKED FOR GITHUB DISKUSSION} for ghcr.io/renovatebot/renovate:36.37.0@sha256:{{MASKED FOR GITHUB DISKUSSION} with digest ghcr.io/renovatebot/renovate@sha256:{{MASKED FOR GITHUB DISKUSSION} ...
$ renovate $RENOVATE_EXTRA_FLAGS
 INFO: Autodiscovered repositories
       "length": 3,
       "repositories": [
         "{{MASKED FOR GITHUB DISKUSSION}/wordpress",
         "{{MASKED FOR GITHUB DISKUSSION}/wordpress",
         "{{MASKED FOR GITHUB DISKUSSION}/wordpress"
       ]
 INFO: Repository started (repository={{MASKED FOR GITHUB DISKUSSION}/wordpress)
       "renovateVersion": "36.37.0"
 INFO: Dependency extraction complete (repository={{MASKED FOR GITHUB DISKUSSION}/wordpress, baseBranch=master)
       "stats": {
         "managers": {
           "composer": {"fileCount": 1, "depCount": 34},
           "npm": {"fileCount": 1, "depCount": 1}
         },
         "total": {"fileCount": 2, "depCount": 35}
       }
```

Is this what you mean?
I included the first repository which is the repository where we need the ACF_PRO_KEY variable.

**Reply by rarkins** [reply]
You need to run with LOG_LEVEL=debug in env to get debug logs

**Reply by JanVormeister** [reply]
This is up top:
```
DEBUG: Env config
       "config": {
         "hostRules": [
           {"hostType": "github", "matchHost": "github.com", "token": "***********"}
         ],
         "extends": ["github>whitesource/merge-confidence:beta"],
         "repositoryCache": "enabled",
         "baseDir": "/builds/{{MASKED FOR GITHUB DISCUSSION}}/renovate-runner/renovate",
         "logFile": "renovate-log.ndjson",
         "logFileLevel": "debug",
         "onboardingConfig": {
           "$schema": "https://docs.renovatebot.com/renovate-schema.json",
           "extends": ["config:recommended"]
         },
         "requireConfig": "required",
         "optimizeForDisabled": true,
         "exposeAllEnv": true,
         "platform": "gitlab",
         "endpoint": "https://{{MASKED FOR GITHUB DISCUSSION}}/api/v4",
         "token": "***********",
         "ignorePrAuthor": true
```

Further down at the repository after the composer update error it's:
```
...
DEBUG: rawExec err (repository={{MASKED FOR GITHUB DISCUSSION}}/wordpress, branch=deps/all-changes)
       "err": {
         "cmd": "/bin/sh -c composer update ...
         "env": {
          "COMPOSER_CACHE_DIR": "/builds/{{MASKED FOR GITHUB DISCUSSION}}/renovate-runner/renovate/cache/others/composer",
          "COMPOSER_AUTH": "{\"github-oauth\":{\"github.com\":\"**redacted**\"},\"gitlab-token\":{\"{{MASKED FOR GITHUB DISCUSSION}}.de\":\"**redacted**\"},\"gitlab-domains\":[\"{{MASKED FOR GITHUB DISCUSSION}}.de\"]}",
          ...

        },
...
```
The ACF_PRO_KEY variable is not included here.
If you need any other specific information from the log, let me know. I don't want to post everything here, because it houlds client data -- obviously.

**Reply by rarkins** [reply]
Unfortunately I'm not able to troubleshoot this further with the information given

**Reply by JanVormeister** [reply]
What information can I provide you to go further?

**Reply by rarkins** [reply]
I think it could take hours to figure out what the issue is, and unfortunately do not have any more time to spare for the topic. Usually if people can make a reproduction then we can debug things reasonably quickly, but in this case it seems environmental so not possible to do.

**Reply by JanVormeister** [reply]
Okay, I understand completely. Thanks for your time though 👍 
But just for the record: renovate should have access to the CI/CD variables in a Self Hosted GitLab environment?

**Reply by rarkins** [reply]
I'm not familiar with GitLab's approach to passing all variables by default or not. If you can create a `config.js` for Renovate and add the line `console.log(Object.keys(process.env).sort());` at the top then it will dump out the list of env it can see, which should match what Renovate sees

**Reply by JanVormeister** [reply]
So I created a `config.js` in the runner repo. The `ACF_PRO_KEY` did not appear in the logs. Then I created a `ACF_PRO_KEY` variable in the runner repo, then it appeared in the logs. Yet, renovate wasn't able to download the plugin because of the missing key. 
I'll keep trying and post here if I find out anything new.

**Reply by rarkins** [reply]
You need any env to be in the repo which the runner runs in, and not the individual repos which are "renovated"

**Reply by JanVormeister** [reply]
Thanks again for your feedback! 

Understood. That's what I did with the`ACF_PRO_KEY` variable. I placed it in the project repo as well as in the renovate runner repo. But it didn't work.

But just to clarify. Renovate needs project based variables on runner level? Isn't this quite unhandy? For example, we have different `ACF_PRO_KEY`s for different projects. So renovate should use each key depending on the project is "renovating".
Sure, we could prefix for each project, but that would be rather uncomfortable.

Or did I missunderstand you?

**Reply by rarkins** [reply]
In the runner model, Renovate runs in one pipeline, not every pipeline. It only has access to variables in the one it's running in

**Reply by JanVormeister** [reply]
I see, makes sense! Thanks again for all your help @rarkins 👍
